File: packages/cli/src/repoint.ts
import type { DatabaseSync } from "node:sqlite";
import { isFileUrl, loadBootstrap, pickAccount, setConfig, type Settings } from "./db.js";
import { resolveJmapBase } from "./discover.js";
import { EXIT, emitJson, fail, note, notFound, out, usage, type IoOpts } from "./io.js";
import { JmapClient, type JmapError } from "./jmap.js";

/**
 * `bullmoose repoint [--base <url>]` — move this device's stored server URL
 * without re-authenticating.
 *
 * The gap it closes: `base` is written once, by `login` or `init`, and nothing
 * could write it again. When a deployment retires a hostname the stored value
 * is wrong forever, and the only repairs were re-running `init` with a token
 * the user may no longer hold a copy of, or opening ~/.bullmoose/mail.db with
 * sqlite3. That is the state PR #201's live smoke found: a config still naming
 * `bullmoose-jmap.eric-d-moore.workers.dev`, which now 404s on everything
 * including `/.well-known/jmap`.
 *
 * It keeps the token and the account and rewrites exactly one config row — and
 * it VALIDATES BEFORE WRITING, so a wrong URL leaves the old (even broken) base
 * in place rather than replacing one wrong answer with another.
 *
 * Not folded into `init`, which requires a token: needing the credential again
 * is precisely the thing that made this unrepairable, and a device whose server
 * moved has not lost its credential.
 */

export interface RepointOpts extends IoOpts {
  /** The new base, or a `file://` bootstrap bundle. Omitted → autodiscover. */
  base?: string;
}

export async function cmdRepoint(db: DatabaseSync, settings: Settings, opts: RepointOpts): Promise<void> {
  const base = await resolveTarget(settings, opts);

  if (base === settings.base) {
    note(`already pointed at ${base}`);
    if (opts.json) emitJson({ base, changed: false });
    return;
  }

  // Validate with the credential already held: a base that serves the session
  // resource but rejects this token is not a repoint, it is a new login, and
  // saying so beats storing it and failing on the next command.
  const session = await new JmapClient(base, settings.token).session().catch((err: unknown) => {
    const status = (err as JmapError).httpStatus;
    if (status === 401 || status === 403) {
      fail(
        `${base} is a JMAP server, but it rejected this device's token (HTTP ${status}).\n` +
          `Nothing was changed. A moved deployment keeps your token; a DIFFERENT one does not — ` +
          `if this is a new server, run: bullmoose login <your address> --base ${base}`,
        EXIT.AUTH,
      );
    }
    throw err;
  });

  if (!session.accounts[settings.accountId]) {
    notFound(
      `${base} does not serve account ${settings.accountId} ` +
        `(it has: ${Object.keys(session.accounts).join(", ") || "none"}).\n` +
        `Nothing was changed — run \`bullmoose login ${session.username}\` if this is a different deployment.`,
    );
  }
  const accounts = Object.entries(session.accounts).map(([id, a]) => ({
    accountId: id,
    name: (a as { name?: string }).name,
  }));

  const previous = settings.base;
  setConfig(db, "base", base);
  setConfig(db, "accounts", JSON.stringify(accounts));
  if (opts.json) {
    emitJson({ base, previousBase: previous, changed: true, accountId: settings.accountId, accounts });
    return;
  }
  out(`repointed: ${previous} -> ${base}`);
  note(`token and account kept (${session.username} / ${settings.accountId})`);
}

/**
 * Where to point. A `file://` base is an operator's bootstrap bundle, exactly as
 * `init` reads it. No base at all asks the network the same question `login`
 * asks — which is the whole point of the bare form: `login`'s answer TODAY,
 * applied to the config `login` wrote months ago.
 */
async function resolveTarget(settings: Settings, opts: RepointOpts): Promise<string> {
  if (isFileUrl(opts.base)) {
    const boot = loadBootstrap(opts.base);
    const fromFile = boot.base ?? boot.url;
    if (!fromFile) usage(`bootstrap bundle names no base: ${opts.base}`);
    return fromFile;
  }
  if (opts.base) return opts.base;

  // The stored address is the input, so a device configured from a bundle that
  // carried no address is told to name the URL rather than handed a DNS error
  // about the empty string.
  const address = pickAccount(settings).address ?? settings.accounts.find((a) => a.address)?.address;
  if (!address) usage("bullmoose repoint --base <url>  (no stored address to autodiscover from)");
  const found = await resolveJmapBase(address);
  note(
    `discovered ${found.base} (via ${found.via})` +
      (found.redirectedFrom ? ` — ${found.redirectedFrom} redirected the session resource here` : ""),
  );
  return found.base;
}

File: cli-go/internal/cmd/repoint.go
package cmd

// `bullmoose repoint [--base <url>]` — move this device's stored server URL
// without re-authenticating. A port of packages/cli/src/repoint.ts, whose
// header records the gap it closes: `base` is written once by login/init, and
// when a deployment retires a hostname the stored value is wrong forever —
// the state PR #201's live smoke actually found.
//
// The discipline that must survive any idiom shift: **validate before
// writing**. A wrong URL leaves the old (even broken) base in place; a base
// that answers but rejects this token is a NEW LOGIN, said in those words,
// not a repoint; a base that lacks this account is refused with the list it
// does have. Only after all three gates does exactly one config row change.

import (
	"context"
	"encoding/json"
	"strconv"
	"strings"

	"github.com/ericdmoore/bullmoose.cc/cli-go/internal/discover"
	bmio "github.com/ericdmoore/bullmoose.cc/cli-go/internal/io"
	"github.com/ericdmoore/bullmoose.cc/cli-go/internal/jmap"
	"github.com/ericdmoore/bullmoose.cc/cli-go/internal/store"
)

func runRepoint(s *bmio.Streams, argv []string) int {
	return runRepointWith(s, argv, discoverDeps{})
}

func runRepointWith(s *bmio.Streams, argv []string, deps discoverDeps) int {
	a := parse(argv)

	db, err := store.Open(store.DBPath(a.DB))
	if err != nil {
		return die(s, err)
	}
	defer db.Close()
	settings, err := store.RequireSettings(db)
	if err != nil {
		return die(s, err)
	}

	base, code := repointTarget(s, settings, a, deps)
	if code != 0 {
		return code
	}

	if base == settings.Base {
		s.Note("already pointed at " + base)
		if a.JSON {
			if err := s.EmitJSON(map[string]any{"base": base, "changed": false}); err != nil {
				return die(s, err)
			}
		}
		return 0
	}

	// Gate 1+2: the base must serve the session resource AND accept the token
	// this device already holds.
	client := jmap.NewSessionClient(base, settings.Token)
	session, err := client.Session(context.Background())
	if err != nil {
		var se *bmio.ServerError
		if asServerError(err, &se) && (se.HTTPStatus == 401 || se.HTTPStatus == 403) {
			s.Note(base + " is a JMAP server, but it rejected this device's token (HTTP " +
				strconv.Itoa(se.HTTPStatus) + ").\nNothing was changed. A moved deployment keeps your token; " +
				"a DIFFERENT one does not — if this is a new server, run: bullmoose login <your address> --base " + base)
			return 4
		}
		return die(s, err)
	}

	// Gate 3: the account this device is bound to must exist there.
	if !session.HasAccount(settings.AccountID) {
		accounts, _ := session.Accounts()
		var ids []string
		for _, acc := range accounts {
			ids = append(ids, acc.ID)
		}
		have := strings.Join(ids, ", ")
		if have == "" {
			have = "none"
		}
		s.Note(base + " does not serve account " + settings.AccountID + " (it has: " + have + ").\n" +
			"Nothing was changed — run `bullmoose login " + session.Username + "` if this is a different deployment.")
		return 1
	}

	sessionAccounts, err := session.Accounts()
	if err != nil {
		return die(s, err)
	}
	type storedAcc struct {
		AccountID string `json:"accountId"`
		Name      string `json:"name,omitempty"`
	}
	stored := make([]storedAcc, 0, len(sessionAccounts))
	for _, acc := range sessionAccounts {
		stored = append(stored, storedAcc{AccountID: acc.ID, Name: acc.Name})
	}
	blob, err := json.Marshal(stored)
	if err != nil {
		return die(s, err)
	}

	previous := settings.Base
	if err := store.SetConfig(db, "base", base); err != nil {
		return die(s, err)
	}
	if err := store.SetConfig(db, "accounts", string(blob)); err != nil {
		return die(s, err)
	}

	if a.JSON {
		if err := s.EmitJSON(map[string]any{
			"base": base, "previousBase": previous, "changed": true,
			"accountId": settings.AccountID, "accounts": stored,
		}); err != nil {
			return die(s, err)
		}
		return 0
	}
	s.Out("repointed: " + previous + " -> " + base)
	s.Note("token and account kept (" + session.Username + " / " + settings.AccountID + ")")
	return 0
}

// repointTarget is repoint.ts:87 resolveTarget: a file:// bundle, an explicit
// --base, or autodiscovery from the stored address — login's answer TODAY,
// applied to the config login wrote months ago.
func repointTarget(s *bmio.Streams, settings *store.Settings, a args, deps discoverDeps) (string, int) {
	if isFileURL(a.Base) {
		boot, err := loadBootstrap(a.Base)
		if err != nil {
			return "", die(s, err)
		}
		fromFile := firstNonEmpty(boot.Base, boot.URL)
		if fromFile == "" {
			s.Note("usage: bootstrap bundle names no base: " + a.Base)
			return "", 2
		}
		return fromFile, 0
	}
	if a.Base != "" {
		return a.Base, 0
	}

	address := ""
	for _, acc := range settings.Accounts {
		if acc.AccountID == settings.AccountID && acc.Address != "" {
			address = acc.Address
			break
		}
	}
	if address == "" {
		for _, acc := range settings.Accounts {
			if acc.Address != "" {
				address = acc.Address
				break
			}
		}
	}
	if address == "" {
		s.Note("usage: bullmoose repoint --base <url>  (no stored address to autodiscover from)")
		return "", 2
	}
	finder := deps.resolver
	if finder == nil {
		finder = discover.New()
	}
	found, err := finder.Resolve(context.Background(), address)
	if err != nil {
		return "", die(s, err)
	}
	line := "discovered " + found.Base + " (via " + found.Via + ")"
	if found.RedirectedFrom != "" {
		line += " — " + found.RedirectedFrom + " redirected the session resource here"
	}
	s.Note(line)
	return found.Base, 0
}
